canonical-web-and-design/maas-ui
# perf(ui): load machines in larger batches after the first request

## Done
Thinking about this issue: https://discourse.maas.io/t/landing-page-takes-multiple-minutes-to-load-2000-nodes/3302 it seems that the number of websocket round trips is a large part of the issue, especially on high latency networks. This branch reduces the number of round trips by fetching 25 machines then batches of 100. For 2000 machines this drops the round trips from 40 to 11 (on my local setup this reduced the total time to load 2000 machines by nearly half).

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Add a heap of machines, using the dev maas you can do:
```console
make harness
from maasserver.testing.factory import factory
for i in range(0, 2000):  factory.make_Node()
```
- Load the machine list. Notice that the first batch of machines will be 25 then should load in batches of 100.
- As it's loading check that the ui rendering isn't slowed down too much as it loads in each batch.
